src/Microsoft.ML.Core/ComponentModel/AssemblyLoadingUtils.cs

```diff
@@ -6,11 +6,11 @@
 using System;
 using System.IO;
 using System.IO.Compression;
-using System.Linq;
 using System.Reflection;
 
 namespace Microsoft.ML.Runtime
 {
+    [BestFriend]
```

Comment:
I don't believe this is the right change for `AssemblyLoadingUtils`. I purposefully made it as "shared source" between `Maml` and the `ML.Legacy` assemblies, with the hopes that `ML.Legacy` will just go away, so this will only be used in `Maml`.

I don't think it should be in the Core assembly (and especially not with a `[BestFriend]` attribute). I don't want people calling this code at all. `Maml` and `ML.Legacy` need it to keep things working, but I don't want any other places calling it.  I've instructed other internal projects, if they need this code, they can copy it and modify it to meet their needs.

Reply:
Well, people have been ignoring your instructions. 😄 A couple more now use it.

I made `ICommand` and all `ICommand` implementors internal. Things that directly call some of the commands now had to reference them. In the case of `Maml` this was `Legacy` and `ResultProcessor`, *both* of which actually use this `AssemblyLoadingUtils`. So I made `Maml` consider `Legacy` and `ResultProcessor` friends of `Maml`, so that they could keep using it. But of course this results in an error, because they're 

However, because you made this a *source* reference, I then had assemblies referencing each other, which led to name collisions resulting in compiler errors. This was my attempt to untangle the mess. I can't just go back to the way things were because that doesn't compile. I didn't perform this move in a fit of whimsy.

It may be that eventually we can winnow it down, but for right now, I'm not sure what else is better to do.

I'm not sure what you mean about "especially not." Just because something is a best-friend internal doesn't mean we're going to keep it now and forever. Indeed the point of introducing this infrastructure in the first place is so we would be free to *later* whip the internal infrastructure code into a better form, without compromising v1.0. This I think falls pretty squarely into its intended use.

Reply:
> I'm not sure what you mean about "especially not." Just because something is a best-friend internal doesn't mean we're going to keep it now and forever. 

It means more places will feel empowered to call this method, because they won't get a "Don't use this" error. And as more places use it, the harder it will be to remove in the future. (Especially places that aren't in this repo, that are planning on getting IVT the Core assembly)

Can we instead do some other "tricks" to make it not part of the Core assembly? Maybe using `extern alias`?  https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/extern-alias

Or maybe we have compiler directives that change the namespace of this file/class in the different assemblies it is used in?

I'd _really_ like this type **not** to be in the Core assembly.

Reply:
Then I'll mark it obsolete as I did with `IHostEnvironment.GetTempFile`. That will give people a nice little message discouraging them, without requiring that I introduce special build infrastructure just for this one little file, which frankly I don't want to do.

---
In reply to: [234269562](https://github.com/dotnet/machinelearning/pull/1626#discussion_r234269562) [](ancestors = 234269562)